internal/app/machined/pkg/system/services/etcd.go

```diff
@@ -683,10 +683,24 @@ func primaryAndListenAddresses() (primary, listen string, err error) {
 		return "", "", errors.New("no valid unicast IP addresses on any interface")
 	}
 
-	// NOTE: we will later likely want to expose the primary IP selection to the
-	// user or build it with greater flexibility.  For now, this maintains
-	// previous behavior.
-	primary = ips[0].String()
+	if subnet == "" {
+		primary = ips[0].String()
+	}
+
+	if primary == "" {
+		network, err := net.ParseCIDR(subnet)
+		if err != nil {
+			return "", "", fmt.Errorf("failed to parse subnet: %w", err)
+		}
+
+		for _, ip := range ips {
+			if network.Contains(ip) {
+				primary = ip.String()
+
+				break
+			}
+		}
+	}
```

Comment:
I guess we should a have check here that if the `primary` is _still_ `""`, then we need to say that no address matched the subnet